token-syncer/src/token_syncer/waiter.clj

```diff
@@ -0,0 +1,160 @@
+                             :query-params {"include-deleted" "true"})]
+      (when error
+        (log/error error "Error in retrieving token" token "from" cluster-url)
+        (throw error))
+      (log/info "Loading" token "on" cluster-url "responded with status" status
+                (select-keys headers ["content-type" "etag" "x-cid"]))
+      (let [last-modified-etag (get headers "etag")]
+        {:description (->> body
+                           async/<!!
+                           try-parse-json-data)
+         :headers (extract-relevant-headers headers)
+         :last-modified-etag last-modified-etag
+         :status status}))
+    (catch Exception ex
+      (log/error "Unable to retrieve token" token "from" cluster-url)
+      {:error ex})))
+
+(defn store-token
+  "Stores the token description on a specific cluster."
+  [^HttpClient http-client cluster-url token last-modified-etag token-description]
+  (log/info "Storing token:" token ", soft-delete:" (true? (get token-description "deleted")) "on" cluster-url)
+  (let [{:keys [body error headers status]}
+        (make-http-request http-client (str cluster-url "/token")
+                           :body (json/write-str (assoc token-description :token token))
+                           :headers {"accept" "application/json"
+                                     "if-match" last-modified-etag}
+                           :method http/post
+                           :query-params {"update-mode" "admin"})
+        body-data (when (not error) (async/<!! body))]
+    (when error
+      (throw error))
+    (log/info "Storing" token "on" cluster-url "responded with status" status
+              (select-keys headers ["content-type" "etag" "x-cid"]))
+    (when (or (nil? status)
+              (< status 200)
+              (> status 299))
+      (log/error "Token store failed" body-data)
+      (throw (ex-info "Token store failed"
+                      {:body body-data
+                       :headers (extract-relevant-headers headers)
+                       :status status
+                       :token-data token-description})))
+    {:body (try-parse-json-data body-data :silent false)
+     :headers (extract-relevant-headers headers)
+     :status status}))
+
+(defn hard-delete-token
+  "Hard-delete a token on a specific cluster."
+  [^HttpClient http-client cluster-url token last-modified-etag]
+  (log/info "Hard-delete" token "on" cluster-url)
+  (let [{:keys [body error headers status]}
+        (make-http-request http-client (str cluster-url "/token")
+                           :headers {"accept" "application/json"
+                                     "if-match" last-modified-etag
+                                     "x-waiter-token" token}
+                           :method http/delete
+                           :query-params {"hard-delete" "true"})
+        body-data (when (not error) (async/<!! body))]
+    (when error
```

Comment:
It seems like returning error is more trouble than it's worth? Would it be shorter overall to just throw in `make-http-request` and catch it in the few cases we need to add additional info.

Reply:
Done.